lib/persist-state.c

```diff
@@ -889,26 +890,36 @@ void
 persist_state_alloc_string(PersistState *self, const gchar *persist_name, const gchar *value, gssize len)
 {
   PersistEntryHandle handle;
-  SerializeArchive *sa;
+  SerializeArchive *serializer;
   GString *buf;
+  gsize buflen;
   gboolean success;
   gpointer block;
 
+  gsize lookup_size;
+  guint8 lookup_result_version;
+
   if (len < 0)
     len = strlen(value);
 
-  buf = g_string_sized_new(len + 5);
-  sa = serialize_string_archive_new(buf);
-
-  success = serialize_write_cstring(sa, value, len);
+  buflen = len + 5;
+  buf = g_string_sized_new(buflen);
+  serializer = serialize_string_archive_new(buf);
+  success = serialize_write_cstring(serializer, value, len);
   g_assert(success == TRUE);
+  serialize_archive_free(serializer);
 
-  serialize_archive_free(sa);
+  handle = persist_state_lookup_entry(self, persist_name, &lookup_size, &lookup_result_version);
+
+  if (!handle || lookup_size < buflen)
```

Comment:
We should use `buf->len` instead of `buflen` here, as it's `serialize_write_cstring()` responsibility to append the `GString` buffer, which may end up being larger than our pre-calculated assumption.